lib/ui/painting.dart

```diff
@@ -1607,12 +1654,160 @@ class Path extends NativeFieldWrapperClass2 {
   }
   Path _transform(Float64List matrix4) native 'Path_transform';
+
+  /// Gets the bounding rectangle for this path.
+  Rect getBounds() {
+    final Float32List rect = _getBounds();
+    return new Rect.fromLTRB(rect[0], rect[1], rect[2], rect[3]);
+  }
+  Float32List _getBounds() native 'Path_getBounds';
+
+  /// Sets this path to the result of operation [PathOperation]
+  /// 
+  /// If only one path is specified, this path will be used as the first 
+  /// operand, otherwise path1 will be the first operand and path2 the
+  /// second.
+  /// 
+  /// The resulting path will be constructed from non-overlapping contours. The
+  /// curve order is reduced where possible so that cubics may be turned into
+  /// quadratics, and quadratics maybe turned into lines.
+  static Path combine(PathOperation operation, Path path1, Path path2, {Path orElse()}) {
+    assert(path1 != null);
+    assert(path2 != null);
+
+    final Path path = new Path();
+    if (path._op(path1, path2, operation.index)) {
+      return path;
+    } else if (orElse != null) {
+      return orElse();
+    }
+    throw new InvalidPathOperationException();
+  }
+  bool _op(Path path1, Path path2, int operation) native 'Path_op';
+
+  /// Creates a [PathMetrics] object for this path.
+  PathMetrics computeMetrics({bool forceClosed = false}) {
+    return new PathMetrics._(this, forceClosed);
+  }
+}
+
+/// Exception thrown by [Path.combine] if a failure occurs.
+class InvalidPathOperationException implements Exception {
+  final String message;
+  InvalidPathOperationException([this.message]);
+}
+
+/// Convenience class to return the result of [PathMetric.getTangentForOffset].
+class Tangent {
+  /// Position of the tangent
+  final Offset position;
+  /// Angle of the tangent in radians
+  final double angle;
+
+  const Tangent(this.position, this.angle);
+}
+
+/// An iterable collection of [PathMetric] objects.
+/// 
+/// Each path metric corresponds to a segment of a path.
+/// 
+/// For example, if you call [Path.moveTo] once in the middle of drawing, you
+/// would get two PathMetric objects in this iterable.  Call 
+/// [Path.computeMetrics] to create the iterable.
+class PathMetrics extends collection.IterableBase<PathMetric> {
+  final Iterator<PathMetric> _iterator;
+
+  PathMetrics._(Path path, bool forceClosed) :
+    _iterator = new PathMetricIterator._(new PathMetric._(path, forceClosed));
+
+  @override
+  Iterator<PathMetric> get iterator => _iterator;
+}
+
+/// Tracks iteration from one segment of a path to the next for measurement.
+class PathMetricIterator implements Iterator<PathMetric> {
+  PathMetric _pathMetric;
+  bool _firstTime = true;
+
+  PathMetricIterator._(this._pathMetric);
+
+  @override
+  PathMetric get current => _pathMetric;
+
+  @override
+  bool moveNext() {
+    // PathMetric isn't a normal iterable - it's already initialized to its
+    // first Path.  Should only call _moveNext when done with the first one. 
```

Comment:
this violates the Iterator contract. We should make sure we obey the contact precisely.

Reply:
(which is easy, just have current return `_firstTime ? null : _pathMetric`)